### ISSUE TITLE
[Performance] Pytorch Model converted to ONNX with CUDAProvicd run slower 3x time than Using Pytorch with GPU 

### ISSUE THREAD

**Issue by hoaileba** [microsoft/onnxruntime#17116]
### Describe the issue

i have a trouble with onnx, I have an pytorch model when i run it with GPU , the time consumed is:
![image](https://github.com/microsoft/onnxruntime/assets/56421735/c33bb337-037d-4261-8fec-4e9fa77b99a7)
but when i convert it to ONNX and run with ONNXRuntime with CUDAProvider : 
![image](https://github.com/microsoft/onnxruntime/assets/56421735/47464bba-f22a-4e6a-bf80-8093a3df9201).

I can't know how to fix this thing. Does anyone can help me with this please
I let my code in reproduce and contact 
ONNX file: https://drive.google.com/file/d/1ho0ivJ2T2cg8wBaN04kEDdc_jRXbV8AF/view?usp=sharing
and shape of input and weight in case you need it
![image](https://github.com/microsoft/onnxruntime/assets/56421735/06bfde6f-058d-4910-8d9c-6d3329081cc6)

### To reproduce

model.py
import torch 
import types, gc, os, time, re
from torch.nn import functional as F

class SA(torch.nn.Module):
    def __init__(self,time_mix_k, time_mix_v, time_mix_r, time_first, time_decay, kw, vw, rw, ow):
        super().__init__()
        self.time_mix_k = time_mix_k
        self.time_mix_v = time_mix_v
        self.time_mix_r = time_mix_r
        self.time_first = time_first
        self.time_decay = time_decay
        self.kw = kw
        self.vw = vw
        self.rw = rw
        self.ow = ow

    def forward(self, x, state_1, state_2, state_3, state_4):
        xk = x * self.time_mix_k + state_1.half() * (1 - self.time_mix_k)
        xv = x * self.time_mix_v + state_1.half() * (1 - self.time_mix_v)
        xr = x * self.time_mix_r + state_1.half() * (1 - self.time_mix_r)
        r = torch.sigmoid(self.rw @ xr)
        kk = torch.matmul(self.kw,xk).float()
        vv = torch.matmul(self.vw, xv).float()

        ww = self.time_first + kk
        p = torch.maximum(state_4, ww)
        e1 = torch.exp(state_4 - p)
        e2 = torch.exp(ww - p)
        a = e1 * state_2 + e2 * vv
        b = e1 * state_3 + e2
        ww = state_4 + self.time_decay
        p = torch.maximum(ww, kk)
        e1 = torch.exp(ww - p)
        e2 = torch.exp(kk - p)
        wkv = (a / b).half()
        
        return self.ow @ (r * wkv), x.float() , e1 * state_2 + e2 * vv, e1 * state_3 + e2, p

### Urgency

it is really important to me, i need it to integrate with the other one in my team

### Platform

Linux

### OS Version

22.04.2 LTS

### ONNX Runtime Installation

Released Package

### ONNX Runtime Version or Commit ID

1.12

### ONNX Runtime API

Python

### Architecture

X64

### Execution Provider

CUDA

### Execution Provider Library Version

CUDA 11.4

### Is this a quantized model?

No

**Comment by hoaileba** [comment]
> @hoaileba , are you using IOBinding for ORT? if not, it is not an apple to apple comparison because there is overhead to copy input/output from CPU to GPU. Please refer to this for example: [#10303 (comment)](https://github.com/microsoft/onnxruntime/issues/10303#issuecomment-1260062482)

thank you for your recommendation, i have tried it and time reduce a lots, now current time consumed is around 0.0014-0.0011s, but it still not better than Pytorch, do i miss anythings, please let me know. below is the code i used.

import torch
import onnxruntime
import numpy

x = torch.randn((4096,), requires_grad=False,dtype=torch.float16).to("cuda").contiguous()
state1 = torch.randn((4096,), requires_grad=False).to("cuda")
state2 = torch.randn((4096,), requires_grad=False).to("cuda")
state3 = torch.randn((4096,), requires_grad=False).to("cuda")
state4 = torch.randn((4096,), requires_grad=False).to("cuda")

so = onnxruntime.SessionOptions()
model = onnxruntime.InferenceSession("sa_1.onnx",so,providers=["CUDAExecutionProvider"])

input_name = model.get_inputs()[0].name
st1 = model.get_inputs()[1].name
st2 = model.get_inputs()[2].name
st3 = model.get_inputs()[3].name
st4 = model.get_inputs()[4].name
output_name = model.get_outputs()[0].name
st_out_1 = model.get_outputs()[1].name
st_out_2 = model.get_outputs()[2].name
st_out_3 = model.get_outputs()[3].name
st_out_4 = model.get_outputs()[4].name
io_binding = model.io_binding()


io_binding.bind_input(
    name=input_name,
    device_type='cuda',
    device_id=0,
    element_type=numpy.float16,
    shape=tuple(x.shape),
    buffer_ptr=x.data_ptr(),
   )
io_binding.bind_input(
    name=st1,
    device_type='cuda',
    device_id=0,
    element_type=numpy.float32,
    shape=tuple(state1.shape),
    buffer_ptr=state1.data_ptr(),
   )
io_binding.bind_input(
    name=st2,
    device_type='cuda',
    device_id=0,
    element_type=numpy.float32,
    shape=tuple(state2.shape),
    buffer_ptr=state2.data_ptr(),
    )
io_binding.bind_input(
    name=st3,
    device_type='cuda',
    device_id=0,
    element_type=numpy.float32,
    shape=tuple(state3.shape),
    buffer_ptr=state3.data_ptr(),
    )
io_binding.bind_input(
    name=st4,
    device_type='cuda',
    device_id=0,
    element_type=numpy.float32,
    shape=tuple(state4.shape),
    buffer_ptr=state4.data_ptr(),
    )

io_binding.bind_output(output_name, 'cuda')
io_binding.bind_output(st_out_1, 'cuda')
io_binding.bind_output(st_out_2, 'cuda')
io_binding.bind_output(st_out_3, 'cuda')
io_binding.bind_output(st_out_4, 'cuda')

model.run_with_iobinding(io_binding)
x_, s1,s2,s3,s4 = io_binding.copy_outputs_to_cpu()

print(input_name,output_name)

print(state1.dtype)
import time
start = time.time()
model.run_with_iobinding(io_binding)
x_, s1,s2,s3,s4 = io_binding.copy_outputs_to_cpu()
# print(x_.shape)
end = time.time()
print(end-start)
